File: reverseWords.py
## Given an input string s, reverse the string word by word.
# Assume input string may contain multiple leading, trailing spaces or spaces between two words
# one pass method: 
class Solution:
    # @param s, a string
    # @return a string
    def reverseWords(self, s):
        output, count = '', 0
        for i in range(len(s) - 1, -1, -1):
            if(s[i] == ' '): count = 0
            else:
                count += 1
                if(i == 0 or s[i - 1] == ' '):
                    output += s[i:i + count]; output += ' '
        return output[:-1]

File: Array/reverseWords.java
/*
 * 
 * Reverse Words in a String 
 * Given an input string, reverse the string word by word.
 * For example,
 * Given s = "the sky is blue", return "blue is sky the".
 * 
 */
class Solution {
	
	/** O(n) runtime */ 
	public String reverseWords(String s){
		StringBuilder reversed = new StringBuilder(); 
		int i = s.length() - 1 ;
		int count; 
		while(i >= 0){
			while(i >= 0 && s.charAt(i) == ' ' ) i --; 
			count = 0; 
			while(i >= 0 && s.charAt(i) != ' ') {
				count ++; i --; 
				if(i < 0 || s.charAt(i ) == ' ') reversed.append(' ').append(s.substring(i + 1, i + 1 + count)) ; 
			}
		}
		return reversed.length() == 0 ? "":reversed.toString().substring(1); 
	}
	
	/** Test case */ 
	public static void main(String[] args) {
		
		String s = "  the sky is blue  "; 
		System.out.println(new Solution().reverseWordsI(s));
		
	}

}
